### ISSUE TITLE
Allow option to view child relations via parent

### ISSUE THREAD

**Issue by convissor** [openstreetmap/openstreetmap-website#2772]
When viewing a relation in the osm website, it doesn't render child relations.  (Example: https://www.openstreetmap.org/relation/9973398)

Was told on IRC that it's because some relations have many nested levels, so it would strain OSM's servers.  Could an optional flag be added that would enable rendering child relations?

**Comment by tomhughes** [comment]
Well that would run the risk of defeating the whole object, if the reason is to stop people being able to kill the server?

**Comment by tomhughes** [comment]
I'm not sure your diagnosis is actually correct but it's a bit hard to be sure without more details about what you think is missing.

It fetches the full relation from the API using the `/api/0.6/relation/N/full` call and then uses https://github.com/openstreetmap/leaflet-osm to render it - if you look at that you'll see there is a comment where it ignore ways and relations embedded in a relation but that actually doesn't matter because it doesn't actually turn relations into features at all - it just generates from the separate node and way objects that are in the API response.

I think what may be happening is that the API response include the relation you asked for, and the ways and relations that are part of it (there is certainly more that one relation in the response) but not the ways which are members of the child relations and hence you effectively don't see the child relation.

Only recursing one deep there is entirely deliberate I believe because otherwise the response might be practically unbounded if you've ever seen some of the insane relations people create...

**Comment by convissor** [comment]
Exactly. It's not gathering/rendering the children ways.

Of course, it can be done. Here's the relation fully rendered using Waymarked Trails (https://cycling.waymarkedtrails.org/#route?id=9973398). (Seems that didn't crash anything. :))

To me, it's unfortunate that OSM website doesn't allow showing that directly, because it's an invitation for people to get involved editing.

You have knowledge of how things work around here. My thought was if the recursion was off by default, but could be enabled by users for particular tasks, chances would be low enough to avoid serious problems. Or maybe just allow a few levels of recursion?

**Comment by tomhughes** [comment]
Well waymarked trails is doing a map rendering, so it's working from a database that contains precomputed geometries created with osm2pgsql, not from an API database which has a rather different schema. It also doesn't do the hard work of resolving those relations in real time.

The biggest difference though is that it selects which relations to process, so it will complete ignore many of the biggest and most complex relations.

In a sense there is already a limited recursion option - that is the difference between the regular API call and the `/full` variant being used here. It's just that the depth limit is 0 or 1 and there's no option for anything more.

Certainly more recursion could be allowed but it would require an API change and changed both here and to the cgimap project that implements the production API.

**Comment by mmd-osm** [comment]
You don't need to the API for this task, use Overpass API instead: http://overpass-turbo.eu/s/X2r

**Comment by mmd-osm** [comment]
By the way, this has some overlap with #2348

**Comment by tomhughes** [comment]
You may be right on a technical level, that the overpass API is capable of returning the data but that requires that we either special case this to fetch the data in a different way or fully switch the data browser to use overpass as the source.

It also requires us to add a new dependency on a third party API that we have no control over.

**Comment by mmd-osm** [comment]
Right, as this is more a question of data visualisation rather than editing using the API, I brought up Overpass API here.

I would discourage unconstrained relation resolution for the API, in particular for use cases like the one mentioned here. Looking at the relation 60189 (Russia) with all its nested multi level subareas, this would quickly produce 584MB worth of data. That's a pretty good way to kill your browser.

Limiting resolution to some given level seems a bit arbitrary to me, as it would for sure also depend on tags on the relation or relation member. That's a good way to introduce lots of special logic that is changing all the time.

**Comment by AndrewKvalheim** [comment]
This makes viewing a [superroute](https://wiki.openstreetmap.org/wiki/Tag:type=superroute) especially confusing as nothing is displayed.

**Comment by mvl22** [comment]
> This makes viewing a [superroute](https://wiki.openstreetmap.org/wiki/Tag:type=superroute) especially confusing as nothing is displayed.

I've just run into this problem.

https://www.openstreetmap.org/relation/17185960/

though this site does render it:

https://hiking.waymarkedtrails.org/#route?id=17185960

If there is some concern about overloading the API or the browser, then I would suggest that:

- API retrieval be limited to X Megabytes, and the result shown if returned and below that level;
- If the route cannot be displayed, the interface **actually says so**; a list of the components of the superroute could for instance be stated as a list, which means the user at least gets some sense of what to do next.

Simply showing nothing because of a technical limitation doesn't help the user.

**Comment by 1ec5** [comment]
> I think what may be happening is that the API response include the relation you asked for, and the ways and relations that are part of it (there is certainly more that one relation in the response) but not the ways which are members of the child relations and hence you effectively don't see the child relation.

The /full response for [this route superrelation](https://www.openstreetmap.org/relation/9973398) includes each member relation and its member relations (which happen to be ways). In turn, that relation belongs to [this route superduperrelation](https://www.openstreetmap.org/relation/11486089), whose /full response recurses three levels down to the ways. Neither response includes any way’s nodes. We’d need the nodes to reconstruct a geometry for display.

https://github.com/openstreetmap/openstreetmap-website/blob/3e447d802ccb269dc468340c54f6c5268f7d4ebf/app/controllers/api/relations_controller.rb#L41-L64

> I would discourage unconstrained relation resolution for the API, in particular for use cases like the one mentioned here. Looking at relation 60189 (Russia) with all its nested multi level subareas, this would quickly produce 584MB worth of data. That's a pretty good way to kill your browser.

Any fix for this issue would definitely need to exclude `subarea` members. Even if the API response includes them, the client-side conversion code would need to ignore them anyways, since the user wouldn’t be expecting the map to highlight subareas.

Even if we could eliminate the `subarea` tagging convention from OSM (pretty please), mappers have gone overboard with other elaborate relation structures, such as [all the numbered roads in South Korea](https://www.openstreetmap.org/relation/6818098). At the same time, there are definitely superrelations that would benefit greatly from a map visualization. (This is especially true in OpenHistoricalMap: OpenHistoricalMap/issues#601.) Unless we have a way to gauge the overall data size ahead of time, it should only be available on demand, after the user explicitly asks for it by clicking a button or something.

> If the route cannot be displayed, the interface **actually says so**; a list of the components of the superroute could for instance be stated as a list, which means the user at least gets some sense of what to do next.

I think this is the biggest pain point for both this issue and #3841. A disclaimer could appear on the map, noting that we’ve loaded the current element and some or all of its members are undownloaded ways.
